clients/githubrepo/branches.go

```diff
@@ -266,5 +419,152 @@ func getBranchRefFrom(data *branch) *clients.BranchRef {
+				ret = append(ret, rule)
+				break
+			}
+
+			if match, err := fnmatch.Match(cond, refName); err != nil {
+				return nil, fmt.Errorf("include match error: %w", err)
+			} else if match {
+				ret = append(ret, rule)
+			}
+		}
+	}
+	return ret, nil
+}
+
+func applyRepoRules(branchRef *clients.BranchRef, rules []*repoRuleSet) {
+	falseVal := false
+	for _, r := range rules {
+		adminEnforced := len(r.BypassActors.Nodes) == 0
+		translated := clients.BranchProtectionRule{
+			EnforceAdmins: &adminEnforced,
+		}
+
+		for _, rule := range r.Rules.Nodes {
+			switch rule.Type {
+			case "DELETION":
+				translated.AllowDeletions = &falseVal
+			case "NON_FAST_FORWARD":
+				translated.AllowForcePushes = &falseVal
+			case "PULL_REQUEST":
+				translatePullRequestRepoRule(&translated, rule)
+			case "REQUIRED_STATUS_CHECKS":
+				translateRequiredStatusRepoRule(&translated, rule)
+			}
+		}
+		mergeBranchProtectionRules(&branchRef.BranchProtectionRule, &translated)
+	}
+}
+
+func translatePullRequestRepoRule(base *clients.BranchProtectionRule, rule *repoRule) {
+	if readBoolPtr(rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush) {
+		base.RequiredPullRequestReviews.DismissStaleReviews = rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush
+	}
+	if readBoolPtr(rule.Parameters.PullRequestParameters.RequireCodeOwnerReview) {
+		base.RequiredPullRequestReviews.RequireCodeOwnerReviews = rule.Parameters.PullRequestParameters.RequireCodeOwnerReview
+	}
+	if readBoolPtr(rule.Parameters.PullRequestParameters.RequireLastPushApproval) {
+		base.RequireLastPushApproval = rule.Parameters.PullRequestParameters.RequireLastPushApproval
+	}
+	if reviewerCount := readIntPtr(rule.Parameters.PullRequestParameters.RequiredApprovingReviewCount); reviewerCount > 0 {
+		base.RequiredPullRequestReviews.RequiredApprovingReviewCount = &reviewerCount
+	}
+}
+
+func translateRequiredStatusRepoRule(base *clients.BranchProtectionRule, rule *repoRule) {
+	statusParams := rule.Parameters.StatusCheckParameters
+	if len(statusParams.RequiredStatusChecks) == 0 {
+		return
+	}
+	enabled := true
+	base.CheckRules.RequiresStatusChecks = &enabled
+	base.CheckRules.UpToDateBeforeMerge = statusParams.StrictRequiredStatusChecksPolicy
+	for _, chk := range statusParams.RequiredStatusChecks {
+		if chk.Context == nil {
+			continue
+		}
+		base.CheckRules.Contexts = append(base.CheckRules.Contexts, *chk.Context)
+	}
+}
+
+func mergeBranchProtectionRules(base, translated *clients.BranchProtectionRule) {
+	var downgrade bool
+	upgrade := true
+
+	// FIXME: upgrade/downgrade clauses need to compare every property?
+	if translated.AllowDeletions != nil && !*translated.AllowDeletions {
+		base.AllowDeletions = translated.AllowDeletions
+		upgrade = upgrade && (base.AllowForcePushes == nil || base.AllowForcePushes == translated.AllowForcePushes)
+		downgrade = downgrade || (base.AllowForcePushes != nil && !*base.AllowForcePushes)
```

Comment:
Are all these `AllowForcePushes` a typo here? The parent `if` is dealing with `AllowDeletions`. Same for the if statement below